atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/CachingTransaction.java

```diff
@@ -40,24 +37,29 @@
 import com.palantir.atlasdb.keyvalue.api.TableReference;
 import com.palantir.atlasdb.keyvalue.impl.Cells;
 import com.palantir.atlasdb.transaction.api.Transaction;
+import com.palantir.atlasdb.transaction.api.TransactionFailedException;
+import com.palantir.atlasdb.transaction.service.TransactionService;
+import com.palantir.util.Pair;
 
 public class CachingTransaction extends ForwardingTransaction {
 
     private static final Logger log = LoggerFactory.getLogger(CachingTransaction.class);
+    private static final long DEFAULT_MAX_CACHED_CELLS = 10_000_000;
 
-    final Transaction delegate;
-
-    private final LoadingCache<String, ConcurrentMap<Cell, byte[]>> columnTableCache = CacheBuilder.newBuilder()
-            .softValues()
-            .build(new CacheLoader<String, ConcurrentMap<Cell, byte[]>>() {
-        @Override
-        public ConcurrentMap<Cell, byte[]> load(String key) throws Exception {
-            return Maps.newConcurrentMap();
-        }
-    });
+    private final Transaction delegate;
+    private final Cache<Pair<String, Cell>, byte[]> cellCache;
 
     public CachingTransaction(Transaction delegate) {
+        this(delegate, DEFAULT_MAX_CACHED_CELLS);
+    }
+
+    public CachingTransaction(Transaction delegate, long maxCachedCells) {
         this.delegate = delegate;
+        cellCache = CacheBuilder.newBuilder()
+                .maximumSize(maxCachedCells)
```

Comment:
Should note that the CachingTransaction is only used by your product and it is not currently wired into the default transaction manager creation paths used by other Atlas consumers.